### DISCUSSION TITLE
[Proposal]: Nullable value types in attribute usage

### DISCUSSION THREAD

**Discussion by matkoch** [dotnet/csharplang#4203] [Language Ideas]
# Nullable value types in attribute usage

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
[summary]: #summary

Remove the restrictions of setting nullable value types in attribute usages.

## Motivation
[motivation]: #motivation

Currently, the following code yields a `CS0655`:

```
public class FooAttribute : Attribute
{
    public bool? Boolean { get; set; }
}

[Foo(Boolean = true)] // CS0655
public class Bar { }
```

This is often an annoyance because I have to fully implement the property with a "backing field":

```
public class FooAttribute : Attribute
{
    public bool? Boolean2; // not even a real backing field, since I might need to access from outside of class
    public bool Boolean
    {
        set => Boolean2 = value;
        get => Boolean2.Value ?? false; // defining a getter is mandatory
    }
}
```

## Detailed design
[design]: #detailed-design

N/A

## Drawbacks
[drawbacks]: #drawbacks

The status quo is a drawback.

## Alternatives
[alternatives]: #alternatives

N/A

## Unresolved questions
[unresolved]: #unresolved-questions

**Comment by Joe4evr** [comment]
Part of the problem here is that [nullable value types cannot be emitted as a `const` value](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHAZgAJ9KBhUgb1MpeqoGMB7AO1kuE6cMAfkoAVABZpuAc0oBeSjAQBXOAG5SAXyA===) which is a requirement for setting attribute properties.

**Reply by matkoch** [reply]
Okay. Maybe I'm oversimplifying, but `true`, `false`, `null`, are all constants. So what's stopping from allowing this?

**Reply by HaloFour** [reply]
The attribute metadata is embedded in the assembly as a binary blob that the is interpreted by the runtime at runtime when you look up the custom attributes on a type/member.  The runtime has to be able to understand the types and layout of the data in that binary blob.  This is why even pseudo-constants supported by C#, like `decimal`, are not supported in attributes, the compiler can't influence how the value would be interpreted.

Would be nice if that list of supported types could be expanded.  I don't believe it's changed since 1.0.

**Comment by matkoch** [comment]


**Reply by YairHalberstadt** [reply]
The runtime only allows specific values in attributes. None of them are the null value of `Nullable<T>`. So this would require runtime changes.

**Reply by matkoch** [reply]
Sorry, I accidentally posted as an answer... will post as reply again..

**Comment by svick** [comment]
I don't understand, why do you have to define the attribute with a backing field? Can't you make it a `bool` auto-property?

```c#
public class FooAttribute : Attribute
{
    public bool Boolean { get; set; }
}
```

Even like this, you should still be able to differentiate between `[Foo]`, `[Foo(Boolean = true)]` and `[Foo(Boolean = false)]`.

**Reply by YairHalberstadt** [reply]
@svick How would you distinguish between the 3 at runtime?

**Reply by svick** [reply]
@YairHalberstadt Using [`GetCustomAttributesData`](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.memberinfo.getcustomattributesdata?view=net-5.0). E.g.:

```c#
using System;
using System.Linq;

foreach (var type in new[] { typeof(Nothing), typeof(True), typeof(False) })
{
    var attributeData = type.GetCustomAttributesData().Single();
    Console.WriteLine($"{attributeData}: {(attributeData.NamedArguments.Any() ? attributeData.NamedArguments.Single() : "nothing")}");
}

public class FooAttribute : Attribute
{
    public bool Boolean { get; set; }
}

[Foo]
class Nothing { }

[Foo(Boolean = true)]
class True { }

[Foo(Boolean = false)]
class False { }
```

This prints:

```
[FooAttribute()]: nothing
[FooAttribute(Boolean = True)]: Boolean = True
[FooAttribute(Boolean = False)]: Boolean = False
```

**Reply by matkoch** [reply]
Good idea, but for one, this doesn't work if you have inheritance chains where an intermediate type sets the property in its constructor. It's a workaround, and it will work in some cases.

**Reply by jnm2** [reply]
IIRC it's safer to use GetCustomAttributesData than to allow arbitrary code to execute by getting an actual attribute instance. Metadata-only/reflection-only contexts enforce this.

**Reply by matkoch** [reply]
But that's unrelated to the proposal?

**Reply by jnm2** [reply]
My answer was to "this doesn't work if you have inheritance chains." An inherited class where an intermediate type sets a property in the constructor is in the same bucket and works equally well with GetCustomAttributesData. Whether or not inheritance is involved, the code that uses reflection to understand the attribute needs to recognize what defaults are set by the constructor that is called if it uses GetCustomAttributesData.

**Comment by Joe4evr** [comment]
Another alternative would be to provide a secondary constructor:
```cs
public class FooAttribute : Attribute
{
    public bool? Boolean { get; }

    public FooAttribute() { }
    public FooAttribute(bool flag) => Boolean = flag;
}
```
Though, I admit, this is only feasible when you only have 1 (or *maaaaaybe* 2) of these properties.
